### PR TITLE
PYFW-389: Support MDNS

### PR DESCRIPTION
**Pull request by husigeza**

### CODE REVIEW

**Comment by jirkadev** [comment]
Hi Géza, are these changes relevant only for the development firmware (development firmware api)? Your changes shouldn't be merged into publish branch (which contains production firmware api)?

File: content/firmwareapi/pycom/network/mdns.md [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+This class provides an interface to use the MDNS functionality.
+
+## Quick Usage Example
+
+Example for advertising own services:
+
+```python
+import time
+from network import MDNS
+# As a first step connection to network should be estbalished, e.g. via WLAN
```
Comment (lilycey) [inline content/firmwareapi/pycom/network/mdns.md:17]: First comment: First, connect to the chosen network e.g. via WLAN 

Note for Geza: Do we use WLANs? I thought we used LPWANs?

File: content/firmwareapi/pycom/network/mdns.md [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+MDNS.init()
+# Set the hostname and instance name of this device
+MDNS.set_name(hostname = "my_hostname", instance_name = "my_instance")
+# Add a TCP service to advertise
+MDNS.add_service("_http", MDNS.PROTO_TCP, 80)
+# Add an UDP service to advertise
+MDNS.add_service("_myservice", MDNS.PROTO_UDP, 1234, txt= (("board","esp32"),("u","user"),("p","password")))
+
+
+# Wait some time and remove 1 service
```
Comment (lilycey) [inline content/firmwareapi/pycom/network/mdns.md]: Second comment: Please be specific with the amount of time needed to wait e.g. 'Wait for approximately 10 seconds'  and also what service needs to be removed

File: content/firmwareapi/pycom/network/mdns.md [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+time.sleep(60)
+# Remove a service, it will not be advertised
```
Comment (lilycey) [inline content/firmwareapi/pycom/network/mdns.md]: Third comment: When a service is removed, it will no longer be advertised

File: content/firmwareapi/pycom/network/mdns.md [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+MDNS.remove_service("_http", MDNS.PROTO_TCP)
+
+```
+
+Example for querying:
+
+```python
+import time
+from network import MDNS
+# As a first step connection to network should be estbalished, e.g. via WLAN
```
Comment (lilycey) [inline content/firmwareapi/pycom/network/mdns.md:41]: Fourth comment: First, connect to the chosen network e.g. via WLAN 

Note for Geza: Do we use WLANs? I thought we used LPWANs?

File: content/firmwareapi/pycom/network/mdns.md [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+
+## Methods
+
+#### MDNS.deinit()
+
+Deinitializes the MDNS module and removes all registered services.
+
+#### MDNS.set_name(\*, hostname=None, instance_name=None)
+
+Sets the hostname and instance name of this device to be advertised.
```
Comment (lilycey) [inline content/firmwareapi/pycom/network/mdns.md]: Sixth comment: Sets the hostname and instance name of the device that is going to be advertised

File: content/firmwareapi/pycom/network/mdns.md [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+
+Arguments are:
```
Comment (lilycey) [inline content/firmwareapi/pycom/network/mdns.md]: Seventh comment: The arguments are:

File: content/firmwareapi/pycom/network/mdns.md [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+* `hostname` is the hostname to set
+* `instance_name` is the instance name to set
+
+#### MDNS.add_service(service_type, proto, port, \*, txt)
+
+Adds a service to the MDNS module which will be advertised.
+
+Arguments are:
+
+* `service_type` is the type of the offered service, e.g.: _http, _ftp or can be custom.
```
Comment (lilycey) [inline content/firmwareapi/pycom/network/mdns.md]: Eighth comment: service_type is the type of the offered service, e.g.: _http, _ftp or can be a custom service

File: content/firmwareapi/pycom/network/mdns.md [hunk 8]
```diff
@@ -0,0 +1,147 @@
...
+* `port` is the port number
+* `txt` is the TXT entries to be placed into the advertisement. The TXT entry is a (key, value) tuple, several TXT entries can be included in an advertisement, in that case this must be given as a list of tuples.
+
+#### MDNS.remove_service(service_type, proto)
+
+Removes a service from the MDNS module so it will not be advertised anymore.
+
+Arguments are:
+
+* `service_type` is the type of the service, e.g.: _http, _ftp or can be custom.
```
Comment (lilycey) [inline content/firmwareapi/pycom/network/mdns.md]: Eighth comment: service_type is the type of the offered service, e.g.: _http, _ftp or can be a custom service

File: content/firmwareapi/pycom/network/mdns.md [hunk 9]
```diff
@@ -0,0 +1,147 @@
...
+* `proto` is the Layer 4 protocol (TCP or UDP), can be `MDNS.PROTO_TCP` or `MDNS.PROTO_UDP`
+
+#### MDNS.query(timeout, service_type, proto, \*, instance_name=None)
+
+Performs a query with the given parameters.
+
+Arguments are:
+
+* `timeout` is the timeout value in miliseconds to be waited for the results
+* `service_type` is the type of the service to be queried, e.g.: _http, _ftp or can be custom.
```
Comment (lilycey) [inline content/firmwareapi/pycom/network/mdns.md]: Eighth comment: service_type is the type of the offered service, e.g.: _http, _ftp or can be a custom service

**Comment by husigeza** [comment]
@jirikrepl : this belongs to a feature which has not been released yet.

File: content/firmwareapi/pycom/network/mdns.md [hunk 10]
```diff
@@ -0,0 +1,147 @@
...
+* `proto` is the Layer 4 protocol (TCP or UDP), can be `MDNS.PROTO_TCP` or `MDNS.PROTO_UDP`
+* `instance_name` is the instance name of the service to be queried
+
+If service found, the function returns with List of `MDNS_Query` objects.
```
Comment (lilycey) [inline content/firmwareapi/pycom/network/mdns.md]: Tenth comment: If the service is found then the function returns with a list of '`MDNS_Query` objects.

File: content/firmwareapi/pycom/network/mdns.md [hunk 11]
```diff
@@ -0,0 +1,147 @@
...
+
+## MDNS_Query class
+
+The `MDNS_Query` aggregates all the properties of an entry of a successfull query session:
```
Comment (lilycey) [inline content/firmwareapi/pycom/network/mdns.md]: 11th comment: The `MDNS_Query` aggregates all of the properties of a successful query session entry:

File: content/firmwareapi/pycom/network/mdns.md [hunk 12]
```diff
@@ -0,0 +1,147 @@
...
+* `hostname` is the hostname of the host advertising the service
+* `instance_name` is the instance_name of the service
+* `addr` is the IPv4 address belonging to the service
+* `port` is the port number belonging to the service
+* `txt`is the TXT entries from the advertisement. The TXT entry is a (key, value) tuple, several TXT entries can be included in an advertisement.
```
Comment (lilycey) [inline content/firmwareapi/pycom/network/mdns.md]: 9th comment: txt`is the TXT entries from the advertisement. The TXT entry is a (key, value) tuple, and several TXT entries can be included in an advertisement.

File: content/firmwareapi/pycom/network/mdns.md [hunk 13]
```diff
@@ -0,0 +1,147 @@
...
+from network import MDNS
+# As a first step connection to network should be estbalished, e.g. via WLAN
+
+# Initialize the MDNS module
+MDNS.init()
+
+# Perform a query for 2000 ms
+q = MDNS.query(2000, "_http", MDNS.PROTO_TCP)
+
+# Print out query's result
```
Comment (lilycey) [inline content/firmwareapi/pycom/network/mdns.md]: Fifth comment: Print out the query's result

File: content/firmwareapi/pycom/network/mdns.md [hunk 14]
```diff
@@ -0,0 +1,147 @@
...
+#### MDNS.add_service(service_type, proto, port, \*, txt)
+
+Adds a service to the MDNS module which will be advertised.
+
+Arguments are:
+
+* `service_type` is the type of the offered service, e.g.: _http, _ftp or can be custom.
+* `proto` is the Layer 4 protocol (TCP or UDP), can be `MDNS.PROTO_TCP` or `MDNS.PROTO_UDP`
+* `port` is the port number
+* `txt` is the TXT entries to be placed into the advertisement. The TXT entry is a (key, value) tuple, several TXT entries can be included in an advertisement, in that case this must be given as a list of tuples.
```
Comment (lilycey) [inline content/firmwareapi/pycom/network/mdns.md]: Ninth comment: `txt` is the TXT entries to be placed into the advertisement. The TXT entry is a (key, value) tuple and several TXT entries can be included in an advertisement. In that case, 'txt' must be given as a list of tuples.

File: content/firmwareapi/pycom/network/mdns.md [hunk 15]
```diff
@@ -0,0 +1,147 @@
...
+* `service_type` is the type of the service, e.g.: _http, _ftp or can be custom.
+* `proto` is the Layer 4 protocol (TCP or UDP), can be `MDNS.PROTO_TCP` or `MDNS.PROTO_UDP`
+
+#### MDNS.query(timeout, service_type, proto, \*, instance_name=None)
+
+Performs a query with the given parameters.
+
+Arguments are:
+
+* `timeout` is the timeout value in miliseconds to be waited for the results
```
Comment (lilycey) [inline content/firmwareapi/pycom/network/mdns.md]: Spelling: `timeout` is the timeout value in milliseconds to wait to receive the results

**Review by lilycey — COMMENTED** [review]
A few edits with one minor question (second comment)

File: content/firmwareapi/pycom/network/mdns.md (hunk 1 above)
Comment (husigeza) [reply content/firmwareapi/pycom/network/mdns.md:17]: WLAN means here Wireless Local Area Network, the device can connect to it via Wi-Fi. We also have a module named WLAN for controlling WiFi specific parts.

File: content/firmwareapi/pycom/network/mdns.md (hunk 2 above)
Comment (husigeza) [reply content/firmwareapi/pycom/network/mdns.md]: It is up to the user to decide. The "remove 1 service" means in the next line we remove 1 service, but I agree it can be confusing, I will correct this.

Comment (lilycey) [reply content/firmwareapi/pycom/network/mdns.md]: Okey, so go for 'Set the wait-time and remove the chosen service (shown in the next line)'

File: content/firmwareapi/pycom/network/mdns.md (hunk 4 above)
Comment (husigeza) [reply content/firmwareapi/pycom/network/mdns.md:41]: Answered above.

File: content/firmwareapi/pycom/network/mdns.md (hunk 5 above)
Comment (husigeza) [reply content/firmwareapi/pycom/network/mdns.md]: I used "this" here because it is the current device where this application is running, not another one.

File: content/firmwareapi/pycom/network/mdns.md (hunk 14 above)
Comment (husigeza) [reply content/firmwareapi/pycom/network/mdns.md]: Fixed.

File: content/firmwareapi/pycom/network/mdns.md (hunk 15 above)
Comment (husigeza) [reply content/firmwareapi/pycom/network/mdns.md]: Fixed.

File: content/firmwareapi/pycom/network/mdns.md (hunk 13 above)
Comment (husigeza) [reply content/firmwareapi/pycom/network/mdns.md]: Fixed.

File: content/firmwareapi/pycom/network/mdns.md (hunk 12 above)
Comment (husigeza) [reply content/firmwareapi/pycom/network/mdns.md]: Fixed.

File: content/firmwareapi/pycom/network/mdns.md (hunk 11 above)
Comment (husigeza) [reply content/firmwareapi/pycom/network/mdns.md]: Fixed.

File: content/firmwareapi/pycom/network/mdns.md (hunk 10 above)
Comment (husigeza) [reply content/firmwareapi/pycom/network/mdns.md]: Fixed.

File: content/firmwareapi/pycom/network/mdns.md (hunk 9 above)
Comment (husigeza) [reply content/firmwareapi/pycom/network/mdns.md]: Fixed.

File: content/firmwareapi/pycom/network/mdns.md (hunk 8 above)
Comment (husigeza) [reply content/firmwareapi/pycom/network/mdns.md]: Fixed.

File: content/firmwareapi/pycom/network/mdns.md (hunk 7 above)
Comment (husigeza) [reply content/firmwareapi/pycom/network/mdns.md]: Fixed.

File: content/firmwareapi/pycom/network/mdns.md (hunk 3 above)
Comment (husigeza) [reply content/firmwareapi/pycom/network/mdns.md]: Fixed.

File: content/firmwareapi/pycom/network/mdns.md (hunk 6 above)
Comment (husigeza) [reply content/firmwareapi/pycom/network/mdns.md]: Fixed.

File: content/firmwareapi/pycom/network/mdns.md [hunk 16]
```diff
@@ -25,10 +25,9 @@ MDNS.add_service("_http", MDNS.PROTO_TCP, 80)
 # Add an UDP service to advertise
 MDNS.add_service("_myservice", MDNS.PROTO_UDP, 1234, txt= (("board","esp32"),("u","user"),("p","password")))
 
-
-# Wait some time and remove 1 service
+# Wait some time
```
Comment (lilycey) [inline content/firmwareapi/pycom/network/mdns.md]: 1 comment - 'Wait some time. You are able to set the amount of time you must wait by altering the code below'

File: content/firmwareapi/pycom/network/mdns.md (hunk 5 above)
Comment (lilycey) [reply content/firmwareapi/pycom/network/mdns.md]: I would still use 'Sets the hostname and instance name of the device that is going to be advertised i.e. the current device that the application is running on' because 'this' reads a bit strangely

**Review by lilycey — COMMENTED** [review]
only two more comments :D

File: content/firmwareapi/pycom/network/mdns.md (hunk 16 above)
Comment (husigeza) [reply content/firmwareapi/pycom/network/mdns.md]: I rephrase it to show what is the purpose of waiting here.

Comment (lilycey) [reply content/firmwareapi/pycom/network/mdns.md]: It's not the purpose that I think folks will have a problem with, it's more that you need to either state approximately how long they might have to wait or how they set that amount of time. 'Wait some time' needs more clarification.

Comment (husigeza) [reply content/firmwareapi/pycom/network/mdns.md]: This line actually does nothing important, it can be completely removed and from functional point of view nothing changes. The example code showing a possible use-case of the new feature, which is:
1. Advertise the services
2. Wait (doesn't matter how long, we cannot advice anything here, this is just an example)
3. Stop advertising 1 of the 2 services

I think the audience of this documentation, who are familiar with Python will not be confused.

**Review by lilycey — APPROVED** [review]
Looks good to me!
